### PR TITLE
add test for flows_reroute_using_default_encap_type feature toogle

### PR DESCRIPTION
**Pull request by andriidovhan**
covers: #2955

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy [hunk 1]
```diff
@@ -112,4 +122,159 @@ class FeatureTogglesV2Spec extends HealthCheckSpecification {
         then: "Able to delete flows"
         [flow, newFlow].each { flowHelperV2.deleteFlow(it.flowId) }
     }
+
+    @Ignore("https://github.com/telstra/open-kilda/issues/2955")
+    @Tags(HARDWARE)
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: why?

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy [hunk 2]
```diff
@@ -112,4 +122,159 @@ class FeatureTogglesV2Spec extends HealthCheckSpecification {
...
+    def "Flow encapsulation type is changed while auto rerouting according to 'flows_reroute_using_default_encap_type' \
+feature toogle"() {
+        given: "A switch pair which supports transit_vlan and vxlan encapsulation types"
+        def swPair = topologyHelper.getAllNeighboringSwitchPairs().find {
+            [it.src, it.dst].every { it.noviflow && !it.wb5164 } && it.paths.size() >= 2
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: I understand what you mean, but this is an 'indirect' search. 'noviflow' switch just means that the manufacturer is 'Noviflow', it is not directly say anything about vxlan. Instead, you should look for 'COPY_FIELD' feature OR look at the switch properties allowed encapsulation type (not sure what is better atm. I guess the second option is better given #2941 is implemented)

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy [hunk 3]
```diff
@@ -112,4 +122,159 @@ class FeatureTogglesV2Spec extends HealthCheckSpecification {
...
+            [it.src, it.dst].every { !it.features.contains(SwitchFeature.NOVIFLOW_COPY_FIELD) } && it.paths.size() >= 2
+        }
+        assumeTrue("Unable to find required switches in topology", swPair as boolean)
+
+        and: "The 'flows_reroute_using_default_encap_type' feature is enabled"
+        def initFeatureToggle = northbound.getFeatureToggles()
+        !initFeatureToggle.flowsRerouteUsingDefaultEncapType && northbound.toggleFeature(FeatureTogglesDto.builder()
+                .flowsRerouteUsingDefaultEncapType(true).build())
+
+        and: "Default encapsulation type is transit_vlan"
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: This is not required step. You can directly create the flow as 'transit vlan' and then change default config to 'vxlan'

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy [hunk 4]
```diff
@@ -112,4 +122,159 @@ class FeatureTogglesV2Spec extends HealthCheckSpecification {
...
+        def initGlobalConfig = northbound.getKildaConfiguration()
+        (initGlobalConfig.flowEncapsulationType == FlowEncapsulationType.TRANSIT_VLAN.toString().toLowerCase()) ?:
+                northbound.updateKildaConfiguration(
+                        new KildaConfigurationDto(flowEncapsulationType: FlowEncapsulationType.TRANSIT_VLAN)
+                )
+
+        and: "A flow with default encapsulation"
+        def initKildaConfig = northbound.getKildaConfiguration()
+        def flow = flowHelperV2.randomFlow(swPair).tap { encapsulationType == null }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: `encapsulationType == null` is not an assignment

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy [hunk 5]
```diff
@@ -112,4 +122,159 @@ class FeatureTogglesV2Spec extends HealthCheckSpecification {
...
+        assumeTrue("Unable to find required switches in topology", swPair as boolean)
+
+        and: "The 'flows_reroute_using_default_encap_type' feature is enabled"
+        def initFeatureToggle = northbound.getFeatureToggles()
+        !initFeatureToggle.flowsRerouteUsingDefaultEncapType && northbound.toggleFeature(FeatureTogglesDto.builder()
+                .flowsRerouteUsingDefaultEncapType(true).build())
+
+        and: "A flow with default encapsulation"
+        def initKildaConfig = northbound.getKildaConfiguration()
+        def flow = flowHelperV2.randomFlow(swPair).tap { encapsulationType == null }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: `encapsulationType == null` is not an assignment

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy (hunk 1 above)
Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: before I tried to find a noviflow switch
for now it is fixed, `HARDWARE` tag is removed

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy (hunk 2 above)
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: On the other hand, _feature_ is more of a background property available only in database which can be changed, i.e. has no 'contract', while switch props are part of API. Also, I'm wondering if we can actually build a vxlan flow on a switch, which has COPY_FIELD but doesn't have VXLAN encap type enabled. Can you check that?

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: I will check it
I believe that system will not create a vxlan flow

Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: System looks on switch properties, not on switch features. you should do the same

Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: Switch can have COPY_FIELD feature, but have a turned off VXLAN encap type property. In this case this switch is considered as a non-vxlan. Your test does not account for this situation

Comment (andriidovhan) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/toggles/FeatureTogglesV2Spec.groovy]: thx, fixed

**Pull request merged**